front_end/src/backend/uploads.js

```diff
@@ -0,0 +1,149 @@
+     * @throws {Error} If bucketName is not one of the enum values in REGIONS_MAP defined in common.js.
+     * 
+     * @param {string} fileName 
+     * @param {string} bucketName 
+     * @returns {number} Returns -0.001 if upload fails - so that a sequence of benchmarks can continue running even if one fails. 
+     */
+    async getDurationOfUpload(fileName, bucketName) {
+        if (!(fileName in FILESIZE_BYTES) || !(fileName in FILESIZE_MIB)) {
+            let errorMessage = `Invalid File Name: '${fileName}'. File names must be any of "2mib.txt", "64mib.txt" or "256mib.txt"`
+            throw new Error(errorMessage);
+        }
+        if (!(bucketName in REGIONS_MAP)) {
+            let errorMessage = `Invalid Bucket Name: '${bucketName}'. Bucket must be a supported Google Cloud Storage Region Name. View https://cloud.google.com/storage/docs/locations for more information.`
+            throw new Error(errorMessage);
+        }
+
+        const fullBucketName = `gcsrbpa-upload-${bucketName}`;
+
+        const headers = {
+            'Content-Type': 'text/plain',
+            'Content-Md5': MD5_SUMS[fileName],
+            'x-goog-content-length-range': `${FILESIZE_BYTES[fileName]},${FILESIZE_BYTES[fileName]}`,
+        };
+
+        const url = await this.getSignedURL(fileName, fullBucketName);
+
+        const axiosOptions = {
+            method: 'PUT',
+            url: url,
+            data: FILE_CONTENTS[fileName],
+            headers: headers,
+            maxContentLength: Infinity,
+            maxBodyLength: Infinity
+        };
+
+        let timeTakenMilliSeconds = -1;
+        try {
+            let start = performance.now();
+            await axios(axiosOptions)
+            timeTakenMilliSeconds = performance.now() - start;
+        } catch (e) {
+            console.log(e);
+        }
+
+        return timeTakenMilliSeconds / 1000;
+    }
+
+    /**
+     * Runs an upload benchmark and returns an Array of an Object keys 'bucketName', 'location', 'fileName', 
+     * 'timeTaken', 'fileSizeBytes', 'speedBps', 'speedMiBps'. 
+     * All numerical results are rounded to three decimal places.
+     * All Object values are strings.
+     * 
+     * @param {string} fileName 
+     * @param {string} bucketName 
+     * @returns {Object.<string, string>[]}
+     */
+    async benchmarkSingleUpload(fileName, bucketName) {
+        const timeTaken = await this.getDurationOfUpload(fileName, bucketName);
+
+        let fileSizeBytes = FILESIZE_BYTES[fileName] || fileName;
+        let fileSizeMiB = FILESIZE_MIB[fileName] || fileName;
+        let location = REGIONS_MAP[bucketName] || bucketName;
+
+        const speedBps = (fileSizeBytes / timeTaken) || -1;
+        const speedMiBps = (fileSizeMiB / timeTaken) || -1;
+
+        let result = [{
+            'bucketName': bucketName,
+            'location': location,
+            'fileName': fileName,
+            'timeTaken': timeTaken.toFixed(3),
```

Comment:
maybe define `3` as a constant above and use that everywhere through this file? this way there are fewer magic numbers. same for `-1`. and by giving it a name, it also becomes more clear what these numbers are